Documentation/bpf.rst

```diff
@@ -3354,7 +3354,376 @@ the toolchain section under the respective tools.
+  and parses the packet to store metadata about the packet. This representation
+  is known as the ``sk_buff``. This structure is then exposed in the BPF input
+  context so that BPF programs from the tc ingress layer can use the metadata that
+  the stack extracts from the packet. This can be useful, but comes with an
+  associated cost of the stack performing this allocation and metadata extraction,
+  and handling the packet until it hits the tc hook. By definition, the ``xdp_buff``
+  doesn't have access to this metadata because the XDP hook is called before
+  this work is done. This is a significant contributor to the performance
+  difference between the XDP and tc hooks.
+
+  Therefore, BPF programs attached to the tc BPF hook can, for instance, read or
+  write the skb's ``mark``, ``pkt_type``, ``protocol``, ``priority``,
+  ``queue_mapping``, ``napi_id``, ``cb[]`` array, ``hash``, ``tc_classid`` or
+  ``tc_index``, vlan metadata, the XDP transferred custom metadata and various
+  other information. All members of the ``struct __sk_buff`` BPF context used
+  in tc BPF are defined in the ``linux/bpf.h`` system header.
+
+  Generally, the ``sk_buff`` is of a completely different nature than
+  ``xdp_buff`` where both come with advantages and disadvantages. For example,
+  the ``sk_buff`` case has the advantage that it is rather straight forward to
+  mangle its associated metadata, however, it also contains a lot of protocol
+  specific information (e.g. GSO related state) which makes it difficult to
+  simply switch protocols by solely rewriting the packet data. This is due to
+  the stack processing the packet based on the metadata rather than having the
+  cost of accessing the packet contents each time. Thus, additional conversion
+  is required from BPF helper functions taking care that ``sk_buff`` internals
+  are properly converted as well. The ``xdp_buff`` case however does not
+  face such issues since it comes at such an early stage where the kernel
+  has not even allocated an ``sk_buff`` yet, thus packet rewrites of any
+  kind can be realized trivially. However, the ``xdp_buff`` case has the
+  disadvantage that ``sk_buff`` metadata is not available for mangling
+  at this time. The latter is overcome by passing custom metadata from
+  XDP BPF to tc BPF, though. In this way, the limitations of each program
+  type can be overcome by operating complementary programs of both types
+  as the use case requires.
+
+..
+
+* Compared to XDP, tc BPF programs can be triggered out of ingress and also
+  egress points in the networking data path as opposed to ingress only in case
+  of XDP.
+
+  The two hook points ``sch_handle_ingress()`` and ``sch_handle_egress()`` in
+  the kernel are triggered out of ``__netif_receive_skb_core()`` and
+  ``__dev_queue_xmit()``, respectively. The latter two are the main receive
+  and transmit functions in the data path that, taking XDP aside, are triggered
```

Comment:
"functions in the data path that, along with XDP, are triggered" although it's clear what your saying anyway.

Reply:
`along with XDP` would not be fully correct. XDP doesn't exist on egress. And XDP ingress could drop packets where they are never seen on tc ingress. Meaning, if you take XDP aside for a moment, then every packet passes through tc ingress and egress hook.

Reply:
Good point, done.